### PR TITLE
balsa: validate trailers even if not enabled.

### PR DESCRIPTION
**Pull request by alyssawilk**
Cloned from https://github.com/envoyproxy/envoy/pull/28958

This brings BalsaParser trailer validation (format and characters) in alignment with http-parser behavior. The downside is that this change makes BalsaParser enforce the header size limit (both on individual lines and on total size of trailer block) even when trailers are not enabled, whereas http-parser does not restrict size at all when trailers are not enabled.

Tracking issue: https://github.com/envoyproxy/envoy/issues/21245

Commit Message: [balsa] Validate trailers even if not enabled.
Additional Description:
Risk Level: low, changed code protected by existing default-false runtime flag.
Testing: //test/common/http/http1:codec_impl_test
Release Notes: n/a
Platform Specific Features: n/a
Runtime guard: envoy.reloadable_features.http1_use_balsa_parser
co-authored-by: bnc@google.com

### CODE REVIEW

File: test/common/http/http1/codec_impl_test.cc [hunk 1]
```diff
@@ -3317,43 +3255,43 @@ TEST_P(Http1ClientConnectionImplTest, LowWatermarkDuringClose) {
 TEST_P(Http1ServerConnectionImplTest, LargeTrailersRejected) {
   // Default limit of 60 KiB
   std::string long_string = "big: " + std::string(60 * 1024, 'q') + "\r\n\r\n\r\n";
-  testTrailersExceedLimit(long_string, "http/1.1 protocol error: trailers size exceeds limit",
+  testTrailersExceedLimit(long_string, "http/1.1 protocol error: trailers size exceeds limit", true,
```
Comment (ravenblackx) [inline test/common/http/http1/codec_impl_test.cc:3258]: Probably all of these false/true could do with `/* fieldname = */` comments to make the test readable.

File: test/common/http/http1/codec_impl_test.cc [hunk 2]
```diff
@@ -3317,43 +3255,43 @@ TEST_P(Http1ClientConnectionImplTest, LowWatermarkDuringClose) {
...
                           true);
 }
 
 // Tests that the default limit for the number of request headers is 100.
 TEST_P(Http1ServerConnectionImplTest, ManyTrailersRejected) {
   // Send a request with 101 headers.
   testTrailersExceedLimit(createHeaderFragment(101) + "\r\n\r\n",
-                          "http/1.1 protocol error: trailers count exceeds limit", true);
+                          "http/1.1 protocol error: trailers count exceeds limit", true, true);
```
Comment (ravenblackx) [inline test/common/http/http1/codec_impl_test.cc:3273]: Comment says headers, test case says trailers, comment says headers, function says trailers, createHeaderFragment says header... Could you make this a little more confused please?

File: test/common/http/http1/codec_impl_test.cc [hunk 3]
```diff
@@ -3317,43 +3255,43 @@ TEST_P(Http1ClientConnectionImplTest, LowWatermarkDuringClose) {
...
 }
 
 TEST_P(Http1ServerConnectionImplTest, LargeTrailersRejectedIgnored) {
   // Default limit of 60 KiB
   std::string long_string = "big: " + std::string(60 * 1024, 'q') + "\r\n\r\n\r\n";
   testTrailersExceedLimit(long_string, "http/1.1 protocol error: trailers size exceeds limit",
-                          false);
+                          false, /* expect_error = */ parser_impl_ == Http1ParserImpl::BalsaParser);
```
Comment (ravenblackx) [inline test/common/http/http1/codec_impl_test.cc:3280]: The test names on these now with param-conditional expect_error could probably be clarified. Even *without* the test validating different things for different params, the test name is already unclear.

**Review by ravenblackx — CHANGES_REQUESTED** [review]

**Review by ravenblackx — APPROVED** [review]

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**